sympy/solvers/solveset.py

```diff
@@ -197,6 +197,11 @@ def invert_real(f_x, y, x, domain=S.Reals):
 def _invert_real(f, g_ys, symbol):
     """Helper function for _invert."""
 
+    # Some inverse functions evaluate to zoo, remove if present
+    fs = FiniteSet(S.ComplexInfinity)
+    if fs.is_subset(g_ys):
+        g_ys -= fs
```

Comment:
I'm not sure this does fix things in a more general way. It removes zoo but loses the reason for removing zoo.

Reply:
Well, the option is to check each and every inverse function that the input will not lead to zoo as you suggested. I cannot really see that keeps the reason better.

Reply:
My point is that once the function class is identified it will in almost all cases call `_invert_real` again with an evaluated version of the inverse function. So rather than checking for every evaluation if the input may yield `zoo`, which is a bit of a hassle and cases will have to be added once identified, it seems more general to check if the input to (the next call of) `_invert_real` is `zoo`. The only possible way that this may not work is if there are some functions that map `zoo` back to reals in a "controlled" way (not just inverting again). It doesn't seem to be any of those cases in the tests so far.

I added a test when I wrote this code to illustrate the point:
`assert solveset(atan(x**2 - y**2)-pi/2, y, S.Reals) == EmptySet()` which currently returns `Intersection({-sqrt(x**2 + zoo), sqrt(x**2 + zoo)}, Reals)`.

Reply:
I think it would be better if `invert_real` only returned sets that are strictly contained in the domain of a function. There are other ways this can go wrong like if if the `subset` test is indeterminate then a zoo case is not removed.

Reply:
> I think it would be better if `invert_real` only returned sets that are strictly contained in the domain of a function

Since it has the `domain` argument, I agree. I think I fixed a problem where the domain was integers but it wasn't return only integers. Those are troublesome bugs because you assume that the routine can be counted on to respect the domain and that you won't have worry about this in the the calling routine.

Reply:
I agree in general. The only other way I might think of is 
```
g_ys = Intersection(g_ys, S.Reals)
```
which will also cover other cases (however, I believe that all the inverse functions return real-valued results for real-valued input, except for those that returns zoo, but I may be wrong). I'm a bit worried that this could explode in certain cases. (`Intersection(ImageSet(..., Intersection(ImageSet(..., Intersection(...)), S.Reals), S.Reals)`) Also, I am not sure how computationally efficient this is, but can give it a try.

Reply:
Will also be interesting see if that deals with: `solveset(I*pi-log(x), x, S.Reals)`, which, if I guess correctly, will have an imaginary intermediate value.

Reply:
I tried to fix this at the intersection-level in #22045.